MCU interface

eirikstor opened the pull request:
Interface between the Onboard Computer (OBC) and microcontroller (MCU). Communication is done over UART.

```serial_rw``` handles ROS subscribers that expects messages in different topics: /mcu_heartbeat, /mcu_arm, /mcu_light_pwm and /pwm.

```SerialCom``` is a class that contains functionality for reading/writing over a serial port (/dev/ttySAC0) and contains the callback functions for the ROS subscribers. 

```crc``` contains cyclic redundancy check functionality

Commit:
Add mcu_interface

This will add functionality to interface with the Giant Gecko.

SerialCom is a class that opens a device in /dev to interface with the MCU.
Callback functions are used to send and receive data when a ROS topic is
published to.

serial_rw.cpp sets up the ROS subscribers and calls ros::spin().

Commit:
Remove symbolic links, add correct files

Commit:
Move header files into include folder

Commit:
Add arming callback function

Create ROS subscriber 'arming_sub'. The callback function when publishing
to this topic is arming_callback. This sends the arming byte the same way
as the heartbeat message does.

Commit:
Correctly add arming function

To arm Manta, publish "ARM" to the /mcu_arm topic. Arming means setting
every PWM signal to 1500 for a finite amount of time (2-3 seconds).

Any other message than "ARM" will disarm the ROV, that is, setting all
PWM signals to 0.

Commit:
Map thruster PWM according to the setup in Manta

For now, create an "alternative_thruster_cmd[]" that maps the PWM signals
according to what thrusters are connected to the ESCs. This must be done
more elegantly in the future

Commit:
Add checksum in alternative command

Commit:
Correctly map thrusters

Note: this is to be done in the MCU in the future

Commit:
Remove thruster remapping

This is done in MCU instead.

Commit:
Introduce light_pwm_callback

Publishing to /mcu_light_pwm will set the corresponding PWM signal

Commit:
Change std:msgs from int32 to int16 in light callback

Commit:
Add mcu_interface node to launchfile. Publish heartbeat to MCU

Commit:
Flush serial input buffer. Read serial data in all callbacks

Commit:
Introduce is_leak() functionality

Commit:
Add uart read on leak message

Commit:
Add leak detection functionality

Also, read buffer size is increased so that larger messages can be read.

Commit:
Rename class to McuInterface

Also, edit CMakelists.txt

aksell committed:
Minor fixes

olesot reviewed (changes requested):
Remove the launch file from the PR. We add complete launch files in a separate PR.

Commit:
Remove bottomside.launch from PR

sindrehan commented on mcu_interface/package.xml:
```diff
@@ -0,0 +1,24 @@
+  <maintainer email="bigbay@todo.todo">bigbay</maintainer>
+
+
+  <license>TODO</license>
```
Set to MIT

sindrehan commented on mcu_interface/package.xml:
```diff
@@ -0,0 +1,24 @@
+  <description>The mcu_interface package</description>
+
+
+  <maintainer email="bigbay@todo.todo">bigbay</maintainer>
```
Fix your e-mail

sindrehan commented on mcu_interface/package.xml:
```diff
@@ -0,0 +1,24 @@
+  <export>
+    <!-- Other tools can request additional information be placed here -->
+
+  </export>
```
Remove the export stuff

Commit:
Remove cout streams. Change ROS_INFO to ROS_DEBUG

sindrehan commented on mcu_interface/src/McuInterface.cpp:
```diff
@@ -0,0 +1,230 @@
+	{
+		m_thruster_cmd[MSG_PAYLOAD_START_INDEX + i*2] = (uint8_t)(msg.positive_width_us[i]  >> 8);
+		m_thruster_cmd[MSG_PAYLOAD_START_INDEX + i*2 + 1] = (uint8_t)(msg.positive_width_us[i]  & 0xFF);
+		printf(" %d ", m_thruster_cmd[i]);
```
If this is necessary use ROS_INFO (or similiar) instead of printf.

sindrehan commented on mcu_interface/src/mcu_interface_node.cpp:
```diff
@@ -0,0 +1,48 @@
+
+int main(int argc, char **argv)
+{
+	const char device[] = "/dev/ttySAC0";
```
The tty device could be parameterised, but it's not strictly necessary.

sindrehan reviewed (dismissed):
Please use spaces instead of tabs, per the style guide. Further, I have added some comments regarding minor things. 

I haven't had time to properly dig through the workings of the code, but if it works and Ole is happy with it we can merge it in.

Commit:
Minor fix ROS_DEBUG vs ROS_INFO

For now, ROS_INFO is published at 1 Hz to notify users that we have
communicatin with MCU

Commit:
Add mail, licence

Commit:
Parameterize device name

Commit:
Add ROS_WARN where applicable.

Also, replace std::cout with ROS_ERROR

Commit:
Minor fix. Increase read buffer size
